### DISCUSSION TITLE
Preprations for merging `development` to `master`

### DISCUSSION THREAD

**Discussion by DRSDavidSoft** [cmderdev/cmder#2878] [Ideas]
@cmderdev/trusted-contributors @Stanzilla @daxgames 

Hi everyone, I was wondering if I could get your agreement on a couple of notes regarding the next version of Cmder:

We should not release another 1.3.x and should instead make a 1.4 release due to the new exciting features by @daxgames (including speed optimization and built-in Windows Terminal support). In addition to this, I propose updating the main screenshot at README.md to reflect the Windows Terminal + Cmder to give Cmder a more fresh look.  I'll also update the website for a more modern, stream-lined and clean look.
There is already a PR over https://github.com/cmderdev/cmder/pull/3081 that demonstrates such an updated hero image.

I'll be happy to hear your opinions; if you agree with the notes above, I'm planning to apply them soon.

The Windows Terminal support development is a great and welcomed new addition to the Cmder, and I'm sure together with the speed optimizations by Dax they're going to hugely improve the user experience of Cmder.

In the meantime, I'd appreciate it if you could please also take a look at the backlog at #2788 and tell me your opinions; once I'm a bit free from my ongoing projects I'm planning to merge the current PRs by Dax and others, and then work on the backlog items as well.

**Comment by daxgames** [comment]
@cmderdev/trusted-contributors @cmderdev/core  Are there compelling reasons I am unaware of that we should not merge `developemnt` into `master` and release v1.4.0 NOW?

I cannot think of any.

We need to release the LARGE CHUNK of features that has been building up for YEARS and get back to small incremental releases.

**Reply by Stanzilla** [reply]
I can't say I'm really involved anymore at this point so I'd leave that judgement to others.

**Reply by daxgames** [reply]
I assumed as much I know it is pretty much me and @DRSDavidSoft but I felt it important to ask everyone.

In my opinion things that need to happen before the next release.

1. Modify the pack process:
  - Make `Windows Terminal` the new default `cmder.zip|7z` archive.
  - Deprecate `ConEmu` to a `cmder-conemu.zip.|7z` archive.
2. Modify the `README.md` to indicate the change above sowing `Windows Terminal` as thebdefault and `ConEmu` as an optional minimally supported terminal.  It does not work most of the time anyway, time to cut the cord.
3. Cut a v2.0.0 release.

After that we go back to incremental feature releases.

**Reply by DRSDavidSoft** [reply]
@daxgames I fully agree with all the points; it's time to fully switch the default terminal from ConEmu to Windows Terminal. I pretty much never open ConEmu these days. Moving forward, the 1.4/2.0 release will be the future of Cmder. 👍🏻

**The feedback from users downloading our releases shows that most people would be comfortable with this change.**

There are still two main reasons why I haven't yet made the move to merge `development` into `master`, *yet*:

* 1️⃣ **Unfinished `master` work:** There are still some code-quality improvements that need to be applied to the existing codebase. There are mistakes in comments, fixes that are still required, and other cleanup work. I've also made the effort to finalize the GitHub Actions CI/CD pipeline and make use of the latest features -- this is especially important now that the AppVeyor days are behind us. I also need to update all the documentation and references to old behaviors and screenshots, as stated above.

  I would prefer to apply and finish these changes on the `master` branch first, for the reason below. However, if my slowness in applying them, or this approach itself, is causing an inconvenience, I apologize and pledge to increase my efforts to complete these changes faster so that we can finally move to the current `development` branch sooner.

* 2️⃣ **Reviewing the `development` branch:** The second reason this merge hasn't happened yet is that, since the introduction of these new features, I haven't really had the time to properly review and approve them. In my mind, I prioritized maintenance work, such as fixing existing issues and improving code readability, over introducing new features (such as the speed improvements).

  There are certainly quite a lot of fixes and improvements already in the `development` branch, as well as the new features that have been merged. However, there are still certain aspects of the `development` branch that would benefit from the same refactoring and code-quality improvements.

  Dax has probably been running the `development` branch for many months now, so he can likely confirm that it's working correctly without a hitch and that there are no significant bugs. Personally, I'm still on the `master` branch and haven't thoroughly tested the `development` branch yet, even though I'm excited for it to become the next official branch.

  There are still aspects of the `development` branch that I'd like to see refactored so they can benefit from the same quality-of-code improvements. For example, the changes to the launcher code requires refactoring to increase readability and maintenance.

All in all, I would kindly request that we hold off on merging `development` into `master`, if possible right now, so that there can be one final 1.2.x or 1.3 release. Such a release would still be useful on older platforms such as Windows 7, where ConEmu is required and Windows Terminal is not available.

If we are to move forward sooner, then perhaps we could rename the old branch to `legacy` and make `development` our new default branch instead.  I'll continue working on that, and will port changes to `development`.

---

P.S. I also vote to rename `master` to `main`; it's a relatively painless switch and is the preferred naming convention these days.

**Reply by daxgames** [reply]
The day we introduced the Windows Terminal is the day I moved off of `master` so it's been years.

Perfect is the enemy of done it's never going to be perfect so it's never going to be done.

We need to release it and fix the things you mentioned incrementally.

**Reply by DRSDavidSoft** [reply]
@daxgames I agree, I'll focus on completing the ongoing PRs within a day or two so that we can merge development.

Tracking the merge here: https://github.com/cmderdev/cmder/pull/2898

**Reply by DRSDavidSoft** [reply]
I'm working on finalizing the current `master` and `development` branches, so they can be cleanly merged.

As discussed, I would like one interim release (the last 1.3.x update) based on `master` before we move on to `development`, since there are quite a lot of code quality improvements made since the last release, and they deserve their own release (also useful for future book keeping).

Before we do that, here are some remaining steps that I need to do:
- Update the Changelog to reflect the changes since the last 1.3.x release
- Provide a "Breaking Changes" / "Important Updates" documentation, highlighting the default terminal change, as well as speed optimizations (the important part is cache invalidation for the init scripts, I still have to ask @daxgames about how it exactly works)
- Work on a few of the remaining `lib_*.cmd` areas that need some light refactoring and touch ups for the 1.3.x release as I prefer these changes to land in `master` rather than `development`.
- Make the "slim" version of Cmder (formerly the Cmder for Windows) be integratable into an already existing Windows Terminal installation as a profile (in fact, this is how I've been using Cmder) instead of bundling a WT with it, for those users who are interested.
- Fix a couple of regressions I inadvertently caused, such as the accidental removal of TC/C shell detection (planned fix for 1.3.x rather than 1.4.x, otherwise it would introduce a regression for those users)
- Explain how the speed optimizations such as `user_init.cmd.template‎` work and provide migration path also for the `.bat` → `.cmd` change

Thank you for bearing with me, I'm aiming for finishing these couple of changes very soon so that we could move on to the new branch as soon as possible.

Lastly, @cmderdev/core please prepare so we can rename `master` to `main`. In the spirit of adopting open-source, I also ask me and dax to be put on the core team, rather than trusted contributures, if that is a possibility.

Thank you for all the great work, dax, chris, and others.
